File: Codeforces/K-th Zero (Easy version)_1600.py
def query(e):
    print("?",1,e,flush = True)
    return e-int(input())

def solve():
    n,t = map(int,input().split())
    k = int(input())
    s,e = 1,n
    while True:
        mid = (s+e)//2
        t = query(mid)
        if t == k and s == e:
            return print("!",s)
        if t>=k:
            e = mid
        else:
            s = mid+1
    print("!",s)
solve()

File: Codeforces/K-th Zero (Easy version)_1600.cpp
#include <bits/stdc++.h>
using namespace std;
#define fastio() ios::sync_with_stdio(false);cin.tie(0); cout.tie(0)


int query(int e)
{
    int sum;
    cout<<"? 1 "<<e<<endl;
    cin>>sum;
    return e-sum;   
}
void solve()
{
    int num,t,k;
    scanf("%d%d%d",&num,&t,&k);
    int s = 1,e = num;
    int mid;
    while (true)
    {
        mid = (s+e)/2;
        t = query(mid);
        if (s == e and t == k)
        {
            cout<<"! "<<s<<endl;
            return;
        }
        if (t>= k)e = mid;
        else s = mid+1;  
    }
}
int main()
{
    fastio();
    solve();
    return 0;
}

/*
1 0 1 0 1 1
*/  